File: src/cmcandy/Java_Language_Answers/_0121.java
package cmcandy.Java_Language_Answers;

/**
 * Created by CHEN on 2020/5/10.
 */
public class _0121 {

    static class Solution {
        public int maxProfit(int[] prices) {

            //贪心或者动态

            //这里是动态

            //数组长度
            int len = prices.length;
            //没有输入则看作没有交易发生，返回0
            if(len==0) return 0;
            //差值最大值
            int max = 0;
            int[] tmp = new int[len];
            tmp[0] = 0;
            for(int i = 1;i<len;i++){
                //如果当前利润大于0，则加上i-(i-1)的两个差的值
                if(tmp[i-1]>0) tmp[i] = prices[i]-prices[i-1]+tmp[i-1];
                else if(tmp[i-1]<=0) tmp[i] = prices[i]-prices[i-1];

                if(max<tmp[i]) max = tmp[i];
            }
            return max;
        }
    }
}

File: src/cmcandy/C_Language_Answers/_0121.cpp
#include <iostream>
#include <vector>
#include <math.h>

using namespace ::std;
class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        if (prices.size() == 0)
            return 0;
        int pro = 0;
        int buy = prices[0];
        for (int i = 1; i < prices.size(); i++)
        {
            //这种情况是倒数
            pro = max(prices[i] - buy, pro);
            if (buy > prices[i])
                buy = prices[i];
        }
        return pro;
    }
};